man/mp4tagcli.1
.\"
.\" mp4tagcli.1
.\"
.\" Copyright 2023-2025 Brad Lanam Pleasant Hill CA USA
.\"
.\" brad.lanam.di / gmail.com
.\"
.TH mp4tagcli 1 "28 Nov 2023"
.SH Name
mp4tagcli \- mp4 tag command line interface
.SH Synopsis
.\" mp4tagcli --version
.\" mp4tagcli --copyfrom in-filename --copyto out-filename
.\" mp4tagcli <filename> --preserve "command-to-run"
.\" mp4tagcli <filename> --clean
.\" mp4tagcli <filename> --duration
.\" mp4tagcli <filename> --asstream
.\" mp4tagcli <filename>
.\" [--binary] [<tag>={|<value>|<filename>}] ...]
.\" [--display <tag> [--dump=<filename>]]
.\" [--freespace <size>]
.\" [<tag>={|<value>|<filename>}] ...]
.B mp4tagcli
\fB\-\-version\fP
.br
.B mp4tagcli
\fB\-\-copyfrom\fP \fIinfile\fP
\fB\-\-copyto\fP \fIoutfile\fP
.br
.B mp4tagcli
\fIfilename\fP
\fB\-\-clean\fP
.br
.br
.B mp4tagcli
\fIfilename\fP
\fB\-\-preserve\fP \fIcommand\-to\-run\fP
.br
.B mp4tagcli
\fIfilename\fP
\fB\-\-duration\fP
.br
.B mp4tagcli
\fIfilename\fP
\fB\-\-asstream\fP
.br
.B mp4tagcli
\fIfilename\fP
[\fB\-\-binary\fP]
[\fB\-\-display\fP \fItag\fP [\fB\-\-dump\fP \fIfilename\fP]]
[\fB\-\-freespace\fP \fIsize\fP]
[\fItag\fP={|\fIvalue\fP|\fIfilename\fP}]
.PP
.SH Description
\fImp4tagcli\fP is used to read and write tags from an MP4 audio or
video file.  Existing tags and existing custom tags will not be lost
when writing or replacing tags.
.PP
\fImp4tagcli\fP can be used as a command line interface to read and write MP4
audio tags, but it is also intended as an example on how to use the
\fIlibmp4tag\fP library, and thus includes other less useful options and
debugging modes.
.SH Overview
.PP
Tags may be specified without the preceding copyright symbol (\[co]).
Any tag that is three characters in length will have the
copyright symbol pre-pended.  This is a function of the
\fImp4tagcli\fP program and not the \fIlibmp4tag\fP library.
.PP
The \fBgnre\fP tag is automatically converted to \fB\[co]gen\fP.  It
is recommended to always use the \fBgen\fP tag.
.PP
When specifying a particular tag, a "\fBnot found\fP" message will be
displayed if the tag is not present (for display) or is not known
(when setting a tag).
.SS Arrays
.PP
Arrays can be specified by suffixing the tag name with a colon
and the index number.  e.g. \fBwrt:1\fP.  Arrays are only implemented
for string data.
.PP
Index values are relative, not absolute.  There will never be an array
with a empty slot.
.SS Cover Images
.PP
There may be more than one cover image present.  If there is more than
one cover image, the subsequent cover images will display a tag name
of \fBcovr:1\fP with the appropriate index value.
.PP
If there is a name associated with the cover image, the tag name will
be displayed as \fBcovr:1:name\fP using the appropriate index value.
.PP
The indexes are only used if there is already an existing cover image
at that index.  Setting a cover image at index 1 in a file with no cover
image present will place the cover image at index 0.
.SS Custom Tags
Custom tags are specified as \-\-\-\-:\fIAPP\fP:\fItagname\fP.  To
prevent the command line parser from recognizing a custom tag as an
option, the end of option specifier \-\- must be added to the
command line before the custom tag.
.PP
e.g. mp4tagcli sample.m4a \-\- \-\-\-\-:BDJ4:DANCE=Waltz
.PP
.SS Displaying Tag Information
.TP
\fBmp4tagcli\fP \fIfilename\fP
Running mp4tagcli without any arguments displays all tags and their
values for \fIfilename\fP.  The duration is also displayed.  Each tag is
output as \fItag\fP\fB=\fP\fIvalue\fP on a separate line.
.TP
\fBmp4tagcli\fP \fIfilename\fP {\fB\-u\fP|\fB\-\-duration\fP}
Prints the duration of the MP4 file without any label.
.TP
\fBmp4tagcli\fP \fIfilename\fP {\fB\-s\fP|\fB\-\-asstream\fP}
Processes the filename as a stream. This is purely for debugging
purposes.
.TP
\fBmp4tagcli\fP \fIfilename\fP \fB{\-d|\-\-display}\fP \fItag\fP
Display \fItag\fP and its value.
.SS Setting Tags
.TP
\fBmp4tagcli\fP \fIfilename\fP [\fB\-\-binary\fP] [\fItag\fP {|\fIvalue\fP|\fIfilename\fP} ...]
Without any value specified, the tag is removed from the MP4 file.
.IP
If a value is specified for a tag, the tag is set to \fIvalue\fP.
.IP
For tags with associated binary data (e.g. cover images), the data for
the tag is read from \fIfilename\fP.
.IP
The \fB\-\-binary\fP option is only used when setting a new tag with
binary data.  It is not necessary to use the \fB\-\-binary\fP option
with cover images or tags with existing binary data.
.IP
Multiple tags may be specified on the command line.
.IP
The \fB\-\-freespace\fP option is used to specify the size of the free
space box to create when the audio file is re-written.
.SS Removing All Tags
.TP
\fBmp4tagcli\fP \fIfilename\fP \fB\-\-clean\fP
Removes all tags from the MP4 file.
.PP
.SS Copying Tags
.TP
\fBmp4tagcli\fP {\fB\-f|\-\-copyfrom\fP} \fIinfile\fP {\fB\-t|\-\-copyto\fP} \fIoutfile\fP
Copies the tags from from \fIinfile\fP to \fIoutfile\fP.  Any existing
tags in \fIoutfile\fP are completely removed.
.PP
.SS Preserving Tags
.TP
\fBmp4tagcli\fP \fIinfile\fP {\fB\-\-preserve\fP} \fIcommand\-to\-run\fP
Preserves the tags in \fIinfile\fP, runs the specfied command, then
restores the tags back to \fIinfile\fP.
.PP
.SH See Also
\fBlibmp4tag\fP(3)
.SH Links
\fBMain\fP: https://sourceforge.net/p/libmp4tag/
.br
\fBWiki\fP: https://sourceforge.net/p/libmp4tag/wiki/
.br
\fBChange Log\fP: https://sourceforge.net/p/libmp4tag/wiki/ChangeLog/
.br
\fBBugs\fP: https://sourceforge.net/p/libmp4tag/tickets/
.SH Copyright
Copyright 2023-2024 Brad Lanam Pleasant Hill
CA USA


ChangeLog.txt
### libmp4tag Change Log

**1.0.2 alpha 2023-8---**

* Bug Fixes
    * fixed genre tags.
    * validate file type
    * mp4tagcli: implement set of binary data.
    * return error codes.
* Changes
    * can now write tags in-place if there is enough space.
* Notes
    * tags cannot be written when there is not enough space.

**1.0.1 alpha 2023-8-17**

* Changes
    * implemented set-tag, delete-tag, clean
    * updated cli tool.
* Notes
    * tags cannot be written

**1.0.0 alpha 2023-8-17**

* Notes
   initial release
    tags cannot be written


mp4tagcli.c
/*
 * Copyright 2023-2025 Brad Lanam Pleasant Hill CA
 */

#include "config.h"

#include <stdio.h>
#include <stdlib.h>
#include <stdint.h>
#include <inttypes.h>
#include <string.h>
#include <getopt.h>
#include <errno.h>

#if __has_include (<windows.h>)
# define WIN32_LEAN_AND_MEAN 1
# include <windows.h>
#endif

#include "libmp4tag.h"

typedef struct {
  int       nargc;
  char      **utf8argv;
} argcopy_t;

static libmp4tag_t * openparse (const char *fname, int dbgflags, int options, int32_t freespacesz);
static libmp4tag_t * openstream_parse (FILE *fh, int dbgflags, int options, int32_t freespacesz);
static void setTagName (const char *tag, char *buff, size_t sz);
static void displayTag (mp4tagpub_t *mp4tagpub);
static void cleanargs (argcopy_t *argcopy);
static size_t clireadcb (char *buff, size_t sz, size_t nmemb, void *udata);
static int cliseekcb (size_t offset, void *udata);

int
main (int argc, char *argv [])
{
  argcopy_t     argcopy;
  libmp4tag_t   *libmp4tag;
  libmp4tagpreserve_t *preservedata;
  mp4tagpub_t   mp4tagpub;
  int           c;
  int           option_index;
  char          tagname [MP4TAG_ID_MAX];
  const         char *infname = NULL;
  const         char *copyto = NULL;
  const         char *preservecmd = NULL;
  const         char *dumpfn = NULL;
  bool          asstream = false;
  bool          clean = false;
  bool          copy = false;
  bool          display = false;
  bool          dump = false;
  bool          duration = false;
  bool          forcebinary = false;
  bool          preserve = false;
  bool          testbin = false;
  bool          write = false;
  int           fnidx = -1;
  int           dbgflags = 0;
  int           options = 0;
  int32_t       freespacesz = 0;
  int           rc = MP4TAG_OK;
  char          *targ;
#if _lib_GetCommandLineW
  wchar_t       **wargv;
  int           targc;
#endif
  FILE          *fh = NULL;          // for openstream test

  static struct option mp4tagcli_options [] = {
    { "asstream",       no_argument,        NULL,   's' },
    { "binary",         no_argument,        NULL,   'b' },
    { "clean",          no_argument,        NULL,   'c' },
    { "copyfrom",       required_argument,  NULL,   'f' },
    { "backup",         no_argument,        NULL,   'k' },
    { "copyto",         required_argument,  NULL,   't' },
    { "debug",          required_argument,  NULL,   'x' },
    { "display",        required_argument,  NULL,   'd' },
    { "dump",           required_argument,  NULL,   'D' },
    { "duration",       no_argument,        NULL,   'u' },
    { "freespace",      required_argument,  NULL,   'F' },
    { "preserve",       required_argument,  NULL,   'P' },
    { "testbin",        no_argument,        NULL,   'B' },
    { "version",        no_argument,        NULL,   'v' },
    { NULL,             0,                  NULL,   0 }
  };

  /* msys2 does not convert argv to utf8 on windows */
  argcopy.nargc = argc;
  argcopy.utf8argv = NULL;
  if (argc > 0) {
    argcopy.utf8argv = malloc (sizeof (char *) * argc);
  }
#if _lib_GetCommandLineW
  wargv = CommandLineToArgvW (GetCommandLineW(), &targc);
  for (int i = 0; i < argc; ++i) {
    argcopy.utf8argv [i] = mp4tag_fromwide (wargv [i]);
  }
  LocalFree (wargv);
#else
  for (int i = 0; i < argc; ++i) {
    argcopy.utf8argv [i] = strdup (argv [i]);
  }
#endif

  *tagname = '\0';

  /* do not specify the 'c' clean short argument */
  while ((c = getopt_long_only (argc, argcopy.utf8argv, "d:D:f:kFt:ux:",
      mp4tagcli_options, &option_index)) != -1) {
    switch (c) {
      case 'b': {
        forcebinary = true;
        break;
      }
      case 'B': {
        testbin = true;
        break;
      }
      case 'c': {
        clean = true;
        break;
      }
      case 'D': {
        dump = true;
        if (optarg != NULL) {
          dumpfn = argcopy.utf8argv [optind - 1];
        }
        break;
      }
      case 'd': {
        display = true;
        if (optarg != NULL) {
          targ = argcopy.utf8argv [optind - 1];
          setTagName (targ, tagname, sizeof (tagname));
        }
        break;
      }
      case 'f': {
        if (optarg != NULL) {
          targ = argcopy.utf8argv [optind - 1];
          infname = targ;
        }
        break;
      }
      case 'F': {
        if (optarg != NULL) {
          targ = argcopy.utf8argv [optind - 1];
          freespacesz = atol (targ);
        }
        break;
      }
      case 'k': {
        options |= MP4TAG_OPTION_KEEP_BACKUP;
        break;
      }
      case 's': {
        asstream = true;
        break;
      }
      case 't': {
        if (optarg != NULL) {
          targ = argcopy.utf8argv [optind - 1];
          copyto = targ;
        }
        break;
      }
      case 'u': {
        duration = true;
        break;
      }
      case 'P': {
        if (optarg != NULL) {
          targ = argcopy.utf8argv [optind - 1];
          preservecmd = targ;
        }
        break;
      }
      case 'v': {
        fprintf (stdout, "mp4tagcli: version %s\n", mp4tag_version ());
        exit (0);
        break;
      }
      case 'x': {
        dbgflags = atoi (optarg);
        break;
      }
      default: {
        break;
      }
    }
  }

  if (infname != NULL && copyto != NULL) {
    copy = true;
  }

  if (! copy) {
    fnidx = optind;
    if (fnidx <= 0 || fnidx >= argc) {
      fprintf (stderr, "no file specified\n");
      cleanargs (&argcopy);
      exit (1);
    }
    infname = argcopy.utf8argv [fnidx];
  }

  if (infname != NULL && preservecmd != NULL) {
    preserve = true;
  }

  if (asstream) {
    fh = fopen (infname, "rb");
    libmp4tag = openstream_parse (fh, dbgflags, options, freespacesz);
  } else {
    libmp4tag = openparse (infname, dbgflags, options, freespacesz);
  }

  if (! asstream && preserve) {
    preservedata = mp4tag_preserve_tags (libmp4tag);
    mp4tag_free (libmp4tag);
    rc = system (preservecmd);
    libmp4tag = openparse (infname, dbgflags, options, freespacesz);
    rc = mp4tag_restore_tags (libmp4tag, preservedata);
    mp4tag_preserve_free (preservedata);
    write = true;
  }

  if (! asstream && copy) {
    preservedata = mp4tag_preserve_tags (libmp4tag);
    mp4tag_free (libmp4tag);
    libmp4tag = openparse (copyto, dbgflags, options, freespacesz);
    mp4tag_restore_tags (libmp4tag, preservedata);
    mp4tag_preserve_free (preservedata);
    write = true;
  }

  if (! asstream && clean && ! copy && ! preserve ) {
    if (mp4tag_clean_tags (libmp4tag) != MP4TAG_OK) {
      fprintf (stderr, "Unable to clean tags (%s)\n", mp4tag_error_str (libmp4tag));
      rc = mp4tag_error (libmp4tag);
    }
    write = true;
  }

  if (rc == MP4TAG_OK && ! asstream && ! clean && ! copy && ! preserve) {
    for (int i = fnidx + 1; i < argc; ++i) {
      char    *tstr;
      char    *tokstr;
      char    *p;

      tstr = strdup (argcopy.utf8argv [i]);
      if (tstr == NULL) {
        continue;
      }

      p = strtok_r (tstr, "=", &tokstr);
      setTagName (tstr, tagname, sizeof (tagname));
      if (p != NULL) {
        p = strtok_r (NULL, "=", &tokstr);
        if (p == NULL) {
          if (mp4tag_delete_tag (libmp4tag, tagname) == MP4TAG_OK) {
            write = true;
          } else {
            fprintf (stderr, "Unable to delete tag: %s (%s)\n", tagname, mp4tag_error_str (libmp4tag));
            rc = mp4tag_error (libmp4tag);
          }
        }
        if (p != NULL) {
          if (testbin) {
            char    *data = NULL;
            size_t  sz;
            int     mp4err;

            data = mp4tag_read_file (p, &sz, &mp4err);
            if (mp4err == MP4TAG_OK && data != NULL) {
              if (mp4tag_set_binary_tag (libmp4tag, tagname, data, sz) == MP4TAG_OK) {
                write = true;
              } else {
                fprintf (stderr, "Unable to set tag: %s (%s)\n", tagname, mp4tag_error_str (libmp4tag));
                rc = mp4tag_error (libmp4tag);
              }
            }
          } else {
            if (mp4tag_set_tag (libmp4tag, tagname, p, forcebinary) == MP4TAG_OK) {
              write = true;
            } else {
              fprintf (stderr, "Unable to set tag: %s (%s)\n", tagname, mp4tag_error_str (libmp4tag));
              rc = mp4tag_error (libmp4tag);
            }
          }
        } /* data is being set for the tag */
      } /* there is a tag name */

      free (tstr);

    } /* for each argument on the command line */
  } /* not clean */

  if (rc == MP4TAG_OK && ! asstream && write) {
    if (mp4tag_write_tags (libmp4tag) != MP4TAG_OK) {
      fprintf (stderr, "Unable to write tags (%s)\n", mp4tag_error_str (libmp4tag));
    }
  }

  if (rc == MP4TAG_OK && display && ! clean && ! copy && ! preserve) {
    int     rc;

    rc = mp4tag_get_tag_by_name (libmp4tag, tagname, &mp4tagpub);
    if (rc != MP4TAG_OK) {
      fprintf (stdout, "%s not found\n", tagname);
    }
    if (rc == MP4TAG_OK) {
      if (! dump) {
        displayTag (&mp4tagpub);
      }
      if (dump &&
          mp4tagpub.binary &&
          mp4tagpub.tag != NULL) {
        FILE    *tfh;

        tfh = fopen (dumpfn, "wb");
        if (tfh != NULL) {
          fwrite (mp4tagpub.data, mp4tagpub.datalen, 1, tfh);
          fclose (tfh);
        }
      }
    }
  }

  if (rc == MP4TAG_OK && ! write && duration && ! preserve) {
    fprintf (stdout, "%" PRId64 "\n", mp4tag_duration (libmp4tag));
  }

  if (rc == MP4TAG_OK && ! write && ! display && ! duration && ! clean && ! preserve) {
    fprintf (stdout, "duration=%" PRId64 "\n", mp4tag_duration (libmp4tag));

    mp4tag_iterate_init (libmp4tag);
    while (mp4tag_iterate (libmp4tag, &mp4tagpub) == MP4TAG_OK) {
      displayTag (&mp4tagpub);
    }
  }

  if (asstream && fh != NULL) {
    fclose (fh);
  }
  mp4tag_free (libmp4tag);
  cleanargs (&argcopy);

  return rc;
}

static void
setTagName (const char *tag, char *buff, size_t sz)
{
  if (strcmp (tag, "art") == 0) {
    tag = "ART";
  }
  if (strcmp (tag, "aart") == 0) {
    tag = "aART";
  }
  if (strlen (tag) == 3 || (strlen (tag) >= 4 && tag [3] == ':')) {
    snprintf (buff, sz, "%s%s", COPYRIGHT_STR, tag);
  } else {
    snprintf (buff, sz, "%s", tag);
  }
}

static void
displayTag (mp4tagpub_t *mp4tagpub)
{
  if (! mp4tagpub->binary &&
      mp4tagpub->tag != NULL &&
      mp4tagpub->data != NULL) {
    if (mp4tagpub->dataidx > 0) {
      fprintf (stdout, "%s:%d=%s\n",
          mp4tagpub->tag, mp4tagpub->dataidx, mp4tagpub->data);
    } else {
      fprintf (stdout, "%s=%s\n", mp4tagpub->tag, mp4tagpub->data);
    }
  }
  if (mp4tagpub->binary &&
      mp4tagpub->tag != NULL) {
    const char  *covertypedisp = "";

    if (strcmp (mp4tagpub->tag, "covr") == 0) {
      covertypedisp = "jpg ";
      if (mp4tagpub->covertype == MP4TAG_COVER_PNG) {
        covertypedisp = "png ";
      }
    }
    if (mp4tagpub->dataidx > 0) {
      fprintf (stdout, "%s:%d=(data: %s%" PRId64 " bytes)\n",
          mp4tagpub->tag, mp4tagpub->dataidx, covertypedisp, (uint64_t) mp4tagpub->datalen);
    } else {
      fprintf (stdout, "%s=(data: %s%" PRId64 " bytes)\n",
          mp4tagpub->tag, covertypedisp, (uint64_t) mp4tagpub->datalen);
    }
    if (mp4tagpub->covername != NULL &&
        *mp4tagpub->covername) {
      /* cover name */
      fprintf (stdout, "%s:%d:name=%s\n",
          mp4tagpub->tag, mp4tagpub->dataidx, mp4tagpub->covername);
    }
  }
}

static libmp4tag_t *
openparse (const char *fname, int dbgflags, int options, int32_t freespacesz)
{
  libmp4tag_t   *libmp4tag = NULL;
  int           mp4error;

  libmp4tag = mp4tag_open (fname, &mp4error);
  if (libmp4tag == NULL) {
    fprintf (stderr, "unable to open %s\n", fname);
    exit (1);
  }

  mp4tag_set_option (libmp4tag, options);
  if (dbgflags != 0) {
    mp4tag_set_debug_flags (libmp4tag, dbgflags);
  }
  if (freespacesz != 0) {
    mp4tag_set_free_space (libmp4tag, freespacesz);
  }

  mp4tag_parse (libmp4tag);
  return libmp4tag;
}

/* this is a very simplistic example using a file handle */
/* a streaming interface would need to provide read and seek callback */
/* functions that work with the user's stream */
static libmp4tag_t *
openstream_parse (FILE *fh, int dbgflags, int options, int32_t freespacesz)
{
  libmp4tag_t   *libmp4tag = NULL;
  int           mp4error;

  libmp4tag = mp4tag_openstream (clireadcb, cliseekcb, fh, 10, &mp4error);
  if (libmp4tag == NULL) {
    fprintf (stderr, "unable to open stream\n");
    exit (1);
  }

  mp4tag_set_option (libmp4tag, options);
  if (dbgflags != 0) {
    mp4tag_set_debug_flags (libmp4tag, dbgflags);
  }
  if (freespacesz != 0) {
    mp4tag_set_free_space (libmp4tag, freespacesz);
  }

  mp4tag_parse (libmp4tag);
  return libmp4tag;
}

static size_t
clireadcb (char *buff, size_t sz, size_t nmemb, void *udata)
{
  FILE    *fh = udata;
  size_t  br;

  br = fread (buff, sz, nmemb, fh);
  return br;
}

static int
cliseekcb (size_t offset, void *udata)
{
  FILE    *fh = udata;
  int     rc;

  rc = mp4tag_fseek (fh, offset, SEEK_CUR);
  return rc;
}

static void
cleanargs (argcopy_t *argcopy)
{
  if (argcopy->utf8argv != NULL) {
    for (int i = 0; i < argcopy->nargc; ++i) {
      if (argcopy->utf8argv [i] != NULL) {
        free (argcopy->utf8argv [i]);
      }
    }

    free (argcopy->utf8argv);
  }
}
